libsolidity/analysis/TypeChecker.cpp

```diff
@@ -596,19 +596,32 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 	if (isStructMemberDeclaration)
 		return false;
 
-	if (auto referenceType = dynamic_cast<ReferenceType const*>(varType))
+	auto compositeType = dynamic_cast<CompositeType const*>(varType);
+	if (compositeType == nullptr)
+		return false;
+
+	bool callDataCheckRequired = true;
+	if (_variable.isLibraryFunctionParameter())
+		callDataCheckRequired = false;
+	else if (_variable.isEventParameter() && _variable.annotation().type->containsNestedMapping())
```

Comment:
Is it still an issue with 2 branches? I think it looks OK now.

Reply:
Now it's even worse because event parameters will never reach the `else` :)

Reply:
I'm confused :). With the following contract we seem to reach the second branch, `else if (_variable.isEventParameter() ....`.

```
contract C {
    event e(mapping(uint => uint) indexed a) anonymous;
}
```

Nevertheless, maybe it is time to invite @ekpyron :).

For certain variables, in addition to the regular location check, we verify if the variable is good for `DataLocation::CallData` also.

Here is the corresponding snippet. I am not sure how accurate it is, do you have an idea?

```
bool callDataCheckRequired = true;
if (
	(_variable.isLibraryFunctionParameter() && compositeType->location() == DataLocation::Storage) ||
	(!_variable.isConstructorParameter() && !_variable.isPublicCallableParameter())
)
	callDataCheckRequired = false;
else if (_variable.isEventParameter() && _variable.isIndexed() && _variable.annotation().type->containsNestedMapping())
{
	solAssert(
		m_errorReporter.errorCount() > 0,
		"Expected \"Type containing a (nested) mapping is not allowed as event parameter type.\""
	);
	callDataCheckRequired = false;
}
```

Reply:
Sorry, you're right. I somehow assumed that all the isXXXParameter functions are mutually exclusive and that `isEventParameter()` implies `!_variable.isConstructorParameter() && !_variable.isPublicCallableParameter()`. In that case the condition is fine.

Continuing discussion from https://github.com/ethereum/solidity/pull/9746#discussion_r494432977: I'm not sure about the indexed/mapping part. First, I think that just being an indexed event parameter should be enough for the checks to be skipped so the `&&` with nested mapping should be an `||`.

As for the nested mappings in events - if they're illegal then I'd just assert that they can't happen here, without complicating the condition.

So my suggestion would be:
```c++
// No need to separately handle nested mappings in events because they're not legal
solAssert(
	!(_variable.isEventParameter() && _variable.annotation().type->containsNestedMapping()),
	"Expected \"Type containing a (nested) mapping is not allowed as event parameter type.\""
);

bool neverStoredInCallData =
	(!_variable.isConstructorParameter() && !_variable.isPublicCallableParameter()) ||
	(_variable.isLibraryFunctionParameter() && compositeType->location() == DataLocation::Storage) ||
	(_variable.isEventParameter() && _variable.isIndexed());
```

Finally, a question to @ekpyron - should (non-indexed) event parameters be checked against the same rules as calldata parameters? They're not stored in calldata, but they're still getting serialized.

Reply:
Also, maybe the assert should be dropped? I'd argue that if we ever decide to allow using mapping as event parameters, they're going to get serialized too and should go through the same checks as all the other parameter types.